### PR TITLE
Fix b/29451090 adding validation and tests for timeout parameters

### PR DESCRIPTION
**Pull request by mpaggi**
Adding code and tests to check the values of the following properties:
adaptor.docContentTimeoutSecs
adaptor.docHeaderTimeoutSecs
adaptor.heartbeatTimeoutSecs
In case the value set in config file is invalid, default value is returned and a warning is logged.

### CODE REVIEW

File: src/com/google/enterprise/adaptor/Config.java [hunk 1]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
     return Long.parseLong(getValue("adaptor.incrementalPollPeriodSecs")) * 1000;
   }
 
+  public long validateTimeout(String property, String defaultValue,
+      boolean allowNeg, boolean allowEmpty) {
+    String secondsAsString = getValue(property).trim();
+    try {
+      if ((null == secondsAsString) || "0".equals(secondsAsString) ||
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: Config.getValue() does not return null.

File: src/com/google/enterprise/adaptor/Config.java [hunk 2]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
     return Long.parseLong(getValue("adaptor.incrementalPollPeriodSecs")) * 1000;
   }
 
+  public long validateTimeout(String property, String defaultValue,
+      boolean allowNeg, boolean allowEmpty) {
```
Comment (PapaJoltOmega) [inline src/com/google/enterprise/adaptor/Config.java]: Why do we have these last two booleans if they are always passed in false?

File: src/com/google/enterprise/adaptor/Config.java [hunk 3]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
...
+        log.log(Level.WARNING, "Invalid value for " + property
+                + ", it has been set to default " + defaultValue + " seconds.");
+      } else {
+          return Long.parseLong(secondsAsString) * 1000;
+      }
+    } catch (NumberFormatException e) {
+      log.log(Level.WARNING, "Invalid value for " + property
+              + ", it has been set to default " + defaultValue + " seconds.");
+    }
+    // For invalid value, returning default
```
Comment (PapaJoltOmega) [inline src/com/google/enterprise/adaptor/Config.java]: Why do we decide to continue with default rather than die?

**Review by PapaJoltOmega — COMMENTED** [review]
Thank you

File: src/com/google/enterprise/adaptor/Config.java [hunk 4]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
...
+      boolean allowNeg, boolean allowEmpty) {
+    String secondsAsString = getValue(property).trim();
+    try {
+      if ((null == secondsAsString) || "0".equals(secondsAsString) ||
+          (("".equals(secondsAsString) && !(allowEmpty))) ||
+          (secondsAsString.startsWith("-") && !(allowNeg))) {
+        log.log(Level.WARNING, "Invalid value for " + property
+                + ", it has been set to default " + defaultValue + " seconds.");
+      } else {
+          return Long.parseLong(secondsAsString) * 1000;
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: The pedantic mission-to-mars programmer would ensure that multiplying by 1000 will not overflow.

File: src/com/google/enterprise/adaptor/Config.java [hunk 5]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
...
+              + ", it has been set to default " + defaultValue + " seconds.");
+    }
+    // For invalid value, returning default
+    // We do not reach this line if value is valid
+    return Long.parseLong(defaultValue) * 1000;
+  }
+
   long getAdaptorDocHeaderTimeoutMillis() {
-    return Long.parseLong(getValue("adaptor.docHeaderTimeoutSecs")) * 1000;
+    return validateTimeout("adaptor.docHeaderTimeoutSecs", "30", false, false);
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: You are restating the default values supplied in the constructor here. I don't like that for 2 reasons:  
1) If the defaults are changed in the future, the change may be missed here.  
2) The default value may have been overridden by the connector using overrideKey(). (The FS adaptor overrides these values)

One solution might be to pull the default values from defaultConfig, but you need to be wary of nulls and computed values (not the case for these).

File: src/com/google/enterprise/adaptor/Config.java [hunk 6]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
     return Long.parseLong(getValue("adaptor.incrementalPollPeriodSecs")) * 1000;
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: you might as well validate this one too.

File: test/com/google/enterprise/adaptor/ConfigTest.java [hunk 7]
```diff
@@ -330,4 +330,48 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
...
+    // docheartbeatTimeoutSecs=0 and docheartbeatTimeoutSecs=-15 are invalid
+    // it should use the default value for adaptor.docHeaderTimeoutSecs
+    // that is 30 seconds.
+    String putInConfig = " gsa.hostname=not_used\n"
+                         + "adaptor.heartbeatTimeoutSecs=";
+    String invalidValuesToVerify[] = {"0", "-15", "", "NotValidValue"};
+    for (int  i = 0; i < invalidValuesToVerify.length; i++) {
+      configFile.setFileContents(putInConfig + invalidValuesToVerify[i]);
+      config.load(configFile);
+      assertEquals(Long.parseLong("30000"),
```
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/ConfigTest.java]: Consider using TimeUnit.SECONDS.toMillis(30).

File: src/com/google/enterprise/adaptor/Config.java [hunk 8]
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
...
 
   long getAdaptorHeartbeatTimeoutMillis() {
-    String secondsAsString = getValue("adaptor.heartbeatTimeoutSecs");
-    if ((null == secondsAsString) || "".equals(secondsAsString.trim())) {
-      secondsAsString = getValue("adaptor.docHeaderTimeoutSecs");
-    }
-    return Long.parseLong(secondsAsString) * 1000;
+    return validateTimeout("adaptor.heartbeatTimeoutSecs", Long.toString(
+                           getAdaptorDocHeaderTimeoutMillis() / 1000),
+                           false, false);
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: The default value of heartbeatTimeoutSecs is "", so allowEmpty should probably be true. But that will still generate a WARNING in the logs.

File: test/com/google/enterprise/adaptor/ConfigTest.java [hunk 9]
```diff
@@ -330,4 +330,48 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
     config.load(configFile);
     assertEquals("\\\t\\\f", config.getValue("slash"));
   }
+
```
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/ConfigTest.java:335]: Add tests that verify that valid timeout config values are returned.

**Review by BrettMichaelJohnson — CHANGES_REQUESTED** [review]

File: src/com/google/enterprise/adaptor/Config.java (hunk 4 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/Config.java]: In fact, I would probably have passed in min and max allowed values (which also does away with allowNeg, which is never true anyway).

File: src/com/google/enterprise/adaptor/Config.java (hunk 3 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/Config.java]: I agree with PJ.  For truly invalid values, throw InvalidConfigurationException.  This also avoids the problems with the defaults below. Although the heartbeatTimeout will still have to fall back to headerTimeout gracefully.

**Comment by BrettMichaelJohnson** [comment]
Michele, we are planning a release of the adaptor library this week.  Please consider finishing this up so that is may be included.

**Comment by PapaJoltOmega** [comment]
Michele, would you mind if Brett took this CL over? Thank you

**Comment by mpaggi** [comment]
Sorry I did not have time to work on this before. Here the changes suggested.
I changed also the tests since 'thrown.expect' does not seems to works in a for cycle.
No problem for me if Brett wants to take over this CL.

File: src/com/google/enterprise/adaptor/Config.java [hunk 10]
```diff
@@ -639,24 +639,42 @@ String getAdaptorFullListingSchedule() {
     return getValue("adaptor.fullListingSchedule");
   }
 
+  public long validateTimeout(String property) {
+    String secondsAsString = getValue(property).trim();
+    if ("adaptor.heartbeatTimeoutSecs".equalsIgnoreCase(property) &&
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/Config.java]: This belongs in getAdaptorHeartbeatTimeoutMillis, not here.

File: test/com/google/enterprise/adaptor/ConfigTest.java [hunk 11]
```diff
@@ -330,4 +332,107 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
...
+        config.getAdaptorHeartbeatTimeoutMillis());
+  }
+
+  @Test
+  public void testPropertiesIncrementalPollPeriodSecs() throws Exception {
+    // incrementalPollPeriodSecs=0 and incrementalPollPeriodSecs=-15 are invalid
+    String putInConfig = " gsa.hostname=not_used\n"
+                         + "adaptor.incrementalPollPeriodSecs=";
+    String invalidValuesToVerify[] = {"0", "-15", "NotValidValue", "",
+                                      Long.toString(Long.MAX_VALUE + 1)};
```
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/ConfigTest.java]: Long.MAX_VALUE + 1 = Long.MIN_VALUE
What you want is (Long.MAX_VALUE/1000)+1 , which will overflow when converted to milliseconds.

File: test/com/google/enterprise/adaptor/ConfigTest.java [hunk 12]
```diff
@@ -330,4 +332,107 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
...
     assertEquals("\\\t\\\f", config.getValue("slash"));
   }
+
+  @Test
+  public void testValidValuesForTimeouts() throws Exception {
+    configFile.setFileContents("gsa.hostname=not_used\n"
+                               + "adaptor.incrementalPollPeriodSecs=1700\n"
+                               + "adaptor.docHeaderTimeoutSecs=50\n"
+                               + "adaptor.docContentTimeoutSecs=190\n"
+                               + "adaptor.heartbeatTimeoutSecs=\n");
```
Comment (BrettMichaelJohnson) [inline test/com/google/enterprise/adaptor/ConfigTest.java]: In this test, explicitly set heartbeat, then add another test that verifies the fallback value.

**Review by BrettMichaelJohnson — CHANGES_REQUESTED** [review]

**Comment by BrettMichaelJohnson** [comment]
PJ,
I took over this CL, and made (admittedly my own) suggested changes.
Please review.

File: src/com/google/enterprise/adaptor/Config.java [hunk 13]
```diff
@@ -639,24 +639,41 @@ String getAdaptorFullListingSchedule() {
...
+    if ("0".equals(secondsAsString) || "".equals(secondsAsString) ||
+        secondsAsString.startsWith("-") ) {
+      throw new InvalidConfigurationException("Invalid value for " + property
+          + ". Zero, empty and negative values are not accepted.");
+    } else {
+      try {
+        return Long.parseLong(secondsAsString) * 1000;
+      } catch (NumberFormatException nfe) {
+        throw new InvalidConfigurationException("Invalid value for "
+            + property + ". Only a numeric value is accepted.");
```
Comment (PapaJoltOmega) [inline src/com/google/enterprise/adaptor/Config.java]: Can we say numeric that also conveys integer?

**Review by PapaJoltOmega — APPROVED** [review]
LGTM.  Thank you

**Review by BrettMichaelJohnson — APPROVED** [review]

**Pull request merged**